File: src/intern_ros_case/src/scripts/movement_node.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import rospy
import math
from geometry_msgs.msg import Twist
from nav_msgs.msg import Odometry
from tf.transformations import euler_from_quaternion

class MovementController:
    def __init__(self):
        rospy.init_node('movement_node', anonymous=True)

        # parameters
        self.linear_speed = rospy.get_param('~linear_speed', 0.2)
        self.angular_speed = rospy.get_param('~angular_speed', 0.3)
        self.move_duration = rospy.get_param('~move_duration', 5.0)

        # Variables
        self.current_yaw = 0.0
        self.current_pose = {'x': 0.0, 'y': 0.0}
        
        # Publisher/Subscriber
        self.velocity_publisher = rospy.Publisher('/cmd_vel', Twist, queue_size=10)
        self.odom_subscriber = rospy.Subscriber('/odom', Odometry, self.odom_callback)
        
        self.rate = rospy.Rate(20)

    # Updates the robot's current position and orientation by processing incoming odometry data.
    def odom_callback(self, msg):
        self.current_pose['x'] = msg.pose.pose.position.x
        self.current_pose['y'] = msg.pose.pose.position.y

        orientation_q = msg.pose.pose.orientation
        quaternion = [orientation_q.x, orientation_q.y, orientation_q.z, orientation_q.w]
        (_, _, yaw) = euler_from_quaternion(quaternion)
        self.current_yaw = yaw

        # Output:
        rospy.loginfo("X: {:.2f} | Y: {:.2f} | Yaw: {:.2f} rad".format(
            self.current_pose['x'], self.current_pose['y'], self.current_yaw))

    # Ensures angular continuity by constraining the angle within the range of -PI to +PI.
    def normalize_angle(self, angle):
        while angle > math.pi: angle -= 2.0 * math.pi
        while angle < -math.pi: angle += 2.0 * math.pi
        return angle

    # Rotates the robot using angular error control until the target orientation is reached.
    def rotate_90_degrees(self):
        vel_msg = Twist()
        target_rad = 1.57   # 90-Degree
        
        start_yaw = self.current_yaw
        target_yaw = self.normalize_angle(start_yaw + target_rad)
        
        rospy.loginfo("Starting Rotation. Target Yaw: {:.2f}".format(target_yaw))

        while not rospy.is_shutdown():
            angle_error = self.normalize_angle(target_yaw - self.current_yaw)
            
            if -0.01 < angle_error < 0.01:
                break
            
            vel_msg.angular.z = self.angular_speed if angle_error > 0 else -self.angular_speed
            self.velocity_publisher.publish(vel_msg)
            self.rate.sleep()

        vel_msg.angular.z = 0
        self.velocity_publisher.publish(vel_msg)
        rospy.loginfo("Rotation completed.")

    # Executes the sequence of tasks sequentially.
    def run_task(self):
        # 1. Forward Motion
        rospy.loginfo("Stage 1: Moving forward...")
        start_time = rospy.get_time()
        vel_msg = Twist()
        
        while rospy.get_time() - start_time < self.move_duration:
            vel_msg.linear.x = self.linear_speed
            self.velocity_publisher.publish(vel_msg)
            self.rate.sleep()

        # 2. Stop
        vel_msg.linear.x = 0
        self.velocity_publisher.publish(vel_msg)
        rospy.sleep(1.0)

        # 3. 90-Degree Rotation 
        rospy.loginfo("Stage 2: Rotating 90 degrees...")
        self.rotate_90_degrees()

        # 4. Stop
        rospy.loginfo("All tasks completed successfully.")
        rospy.spin()

if __name__ == '__main__':
    try:
        controller = MovementController()
        rospy.sleep(1.0) 
        controller.run_task()
    except rospy.ROSInterruptException:
        pass

File: src/intern_ros_case/src/movement_node.cpp
#include "intern_ros_case/movement_node.hpp"

MovementController::MovementController() : rate(20) {
    // parameters
    ros::NodeHandle pnh("~");
    pnh.param<double>("linear_speed", linear_speed, 0.2);
    pnh.param<double>("angular_speed", angular_speed, 0.3);
    pnh.param<double>("move_duration", move_duration, 5.0);

    // Variables
    current_yaw = 0.0;
    current_pose["x"] = 0.0;
    current_pose["y"] = 0.0;

    // Publisher/Subscriber
    velocity_publisher = nh.advertise<geometry_msgs::Twist>("/cmd_vel", 10);
    odom_subscriber = nh.subscribe("/odom", 10, &MovementController::odom_callback, this);
}

void MovementController::odom_callback(const nav_msgs::Odometry::ConstPtr& msg) {
    current_pose["x"] = msg->pose.pose.position.x;
    current_pose["y"] = msg->pose.pose.position.y;

    tf2::Quaternion q(
        msg->pose.pose.orientation.x,
        msg->pose.pose.orientation.y,
        msg->pose.pose.orientation.z,
        msg->pose.pose.orientation.w);
    
    double roll, pitch, yaw;
    tf2::Matrix3x3 m(q);
    m.getRPY(roll, pitch, yaw);
    current_yaw = yaw;

    // Output:
    ROS_INFO("X: %.2f | Y: %.2f | Yaw: %.2f rad", 
             current_pose["x"], current_pose["y"], current_yaw);
}

double MovementController::normalize_angle(double angle) {
    while (angle > M_PI) angle -= 2.0 * M_PI;
    while (angle < -M_PI) angle += 2.0 * M_PI;
    return angle;
}

void MovementController::rotate_90_degrees() {
    geometry_msgs::Twist vel_msg;
    double target_rad = 1.57;   // 90-Degree
    
    double start_yaw = current_yaw;
    double target_yaw = normalize_angle(start_yaw + target_rad);
    
    ROS_INFO("Starting Rotation. Target Yaw: %.2f", target_yaw);

    while (ros::ok()) {
        ros::spinOnce(); 

        double angle_error = normalize_angle(target_yaw - current_yaw);
        
        if (std::abs(angle_error) < 0.01) {
            break;
        }
        
        vel_msg.angular.z = (angle_error > 0) ? angular_speed : -angular_speed;
        velocity_publisher.publish(vel_msg);
        rate.sleep();
    }

    vel_msg.angular.z = 0;
    velocity_publisher.publish(vel_msg);
    ROS_INFO("Rotation completed.");
}

void MovementController::run_task() {
    // 1. Forward Motion
    ROS_INFO("Stage 1: Moving forward...");
    ros::Time start_time = ros::Time::now();
    geometry_msgs::Twist vel_msg;
    
    while (ros::ok() && (ros::Time::now() - start_time).toSec() < move_duration) {
        vel_msg.linear.x = linear_speed;
        velocity_publisher.publish(vel_msg);
        ros::spinOnce();
        rate.sleep();
    }

    // 2. Stop
    vel_msg.linear.x = 0;
    velocity_publisher.publish(vel_msg);
    ros::Duration(1.0).sleep();

    // 3. 90-Degree Rotation 
    ROS_INFO("Stage 2: Rotating 90 degrees...");
    rotate_90_degrees();

    // 4. Stop
    ROS_INFO("All tasks completed successfully.");
    ros::spin();
}

int main(int argc, char** argv) {
    ros::init(argc, argv, "movement_node");
    
    try {
        MovementController controller;
        ros::Duration(1.0).sleep(); 
        controller.run_task();
    }
    catch (ros::Exception& e) {
        ROS_ERROR("ROS Error: %s", e.what());
    }

    return 0;
}